### PR TITLE
Added Order and Order Item. Added endpoints for management

### PR DESCRIPTION
**Pull request by Nikname2303**

### CODE REVIEW

File: src/main/java/com/example/bookshop/dto/order/OrderRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.example.bookshop.dto.order;
+
+import lombok.Data;
+
+@Data
+public class OrderRequestDto {
+    private String shippingAddress;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/dto/order/OrderRequestDto.java:9]: Please validate this field

File: src/main/java/com/example/bookshop/dto/order/OrderUpdateDto.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookshop.dto.order;
+
+import com.example.bookshop.model.Order;
+import lombok.Data;
+
+@Data
+public class OrderUpdateDto {
+    private Order.Status status;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/dto/order/OrderUpdateDto.java:10]: Please validate this field

File: src/main/java/com/example/bookshop/mapper/OrderItemMapper.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import com.example.bookshop.model.OrderItem;
+import java.util.Set;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
+public interface OrderItemMapper {
+    Set<OrderItemResponseDto> toSetDto(Set<OrderItem> orderItems);
+
+    OrderItemResponseDto toDto(OrderItem orderItem);
+
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/mapper/OrderItemMapper.java:16]:
```suggestion
```

File: src/main/java/com/example/bookshop/mapper/OrderMapper.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+public interface OrderMapper {
+    @Mapping(source = "user.id", target = "userId")
+    Set<OrderResponseDto> toSetDto(Set<Order> orders);
+
+    @Mapping(source = "user.id", target = "userId")
+    OrderResponseDto toDto(Order order);
+
+    @Mapping(source = "user.id", target = "userId")
+    OrderResponsePatchDto toPatchDto(Order order);
+
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/mapper/OrderMapper.java]:
```suggestion
```

File: src/main/java/com/example/bookshop/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/model/Order.java:26]: Please don`t forget to add equals, hashCode and to String to entities

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final UserRepository userRepository;
+
+    @Override
+    public Set<OrderResponseDto> getOrders(Long id) {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]:
```suggestion
    public Set<OrderResponseDto> getOrders(Long userId) {
```

Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]:
```suggestion
    public Set<OrderResponseDto> getAll(Long id) {
```

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        Set<Order> orders = orderRepository.findAllByUserId(id);
+        return orderMapper.toSetDto(orders);
+    }
+
+    @Override
+    public Set<OrderItemResponseDto> getAllOrderItemsById(Long userId, Long orderId) {
+        Set<Order> orders = orderRepository.findAllByUserId(userId);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]: What is the reason for retrieving all orders?

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        Order order = new Order();
+        order.setOrderDate(LocalDateTime.now());
+        order.setTotal(BigDecimal.valueOf(0));
+        order.setUser(userRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find user with id: " + id)
+        ));
+        order.setStatus(Order.Status.IN_PROGRESS);
+        order.setShippingAddress(address);
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]: You should find order by id and then update it, not create new one and save it, also check if this order belongs to current user

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public OrderItemResponseDto getOrderItemById(String email, Long orderId, Long itemId) {
+        Order order = orderRepository.getOrderByUserEmail(email);
+        Set<OrderItem> orderItems = order.getOrderItems();
+        for (OrderItem oi : orderItems) {
+            if (oi.getId().equals(itemId)) {
+                return orderItemMapper.toDto(oi);
+            }
+        }
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]: You can write single query to get all necessary information

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public OrderResponsePatchDto updateStatus(String email, Long orderId, Order.Status status) {
+        Order order = orderRepository.getOrderByUserEmail(email);
+        order.setStatus(status);
+        return orderMapper.toPatchDto(order);
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItemById(String email, Long orderId, Long itemId) {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]:
```suggestion
    public OrderItemResponseDto getOrderItemById(Long userId, Long orderId, Long itemId) {
```

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                () -> new EntityNotFoundException("Can't find user with id: " + id)
+        ));
+        order.setStatus(Order.Status.IN_PROGRESS);
+        order.setShippingAddress(address);
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public OrderResponsePatchDto updateStatus(String email, Long orderId, Order.Status status) {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]:
```suggestion
    public OrderResponsePatchDto updateStatus(Long userId, Long orderId, Order.Status status) {
```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job! Left some comments

File: src/main/java/com/example/bookshop/dto/order/OrderUpdateDto.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package com.example.bookshop.dto.order;
+
+import com.example.bookshop.model.Order;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class OrderUpdateDto {
+    @NotBlank
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/dto/order/OrderUpdateDto.java]:
```suggestion
    @NotNull
```

File: src/main/java/com/example/bookshop/model/Order.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    @ManyToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/model/Order.java:44]:
```suggestion
    @CreationTimestamp
    private LocalDateTime orderDate;
```

File: src/main/java/com/example/bookshop/model/Order.java (hunk 5 above)
Comment (d1sam) [reply src/main/java/com/example/bookshop/model/Order.java:26]: please exclude `user`, `orderItems` from `@EqualsAndHashCode`, `@ToString`

File: src/main/java/com/example/bookshop/model/OrderItem.java
Comment (d1sam) [inline src/main/java/com/example/bookshop/model/OrderItem.java:1]: please also exclude related Entities from `@ToString` and `@EqualsAndHashCode`

File: src/main/java/com/example/bookshop/model/OrderItem.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+@Getter
+@Setter
+@EqualsAndHashCode
+@ToString
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetchType = FetchType.LAZY)
```

File: src/main/java/com/example/bookshop/model/OrderItem.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @OneToOne
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/model/OrderItem.java]:
```suggestion
    @OneToOne(fetchType = FetchType.LAZY)
```

File: src/main/java/com/example/bookshop/model/Order.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+@Getter
+@Setter
+@EqualsAndHashCode
+@ToString
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/model/Order.java]:
```suggestion
    @ManyToOne(fetchType = FetchType.LAZY)
```

File: src/main/java/com/example/bookshop/model/OrderItem.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    @Column(nullable = false)
+    private int quantity;
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/model/OrderItem.java]:
```suggestion
    private Integer quantity;
```
I prefer to use the wrapped type as you can tell if the object has been saved/loaded to/from the database whether or not the id value is null.

File: src/main/java/com/example/bookshop/repository/OrderItemRepository.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
+package com.example.bookshop.repository;
+
+import com.example.bookshop.model.OrderItem;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/repository/OrderItemRepository.java:11]: if you add `@Repository` in `OrderRepository` interf, you should add `@Repository` in `OrderItemRepository` for the consistency

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    public Set<OrderItemResponseDto> getAllOrderItemsById(Long userId, Long orderId) {
+        Order order = orderRepository.findByIdAndUserId(orderId, userId).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find order for user by this id: " + userId)
+        );
+        return orderItemMapper.toSetDto(order.getOrderItems());
+    }
+
+    @Override
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java:51]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    public OrderResponsePatchDto updateStatus(Long userId, Long orderId, Order.Status status) {
+        Order order = orderRepository.findById(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find order for user with this id: "
+                        + userId)
+        );
+        order.setStatus(status);
+        return orderMapper.toPatchDto(order);
```
Comment (d1sam) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]:
```suggestion
        order.setStatus(status);
        orderRepository.save(order);
        return orderMapper.toPatchDto(order);
```

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java
Comment (d1sam) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java:1]: what about createOrder?

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but there are some changes, that should be done, according to my comments.

File: src/main/java/com/example/bookshop/controller/OrderController.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Order management", description = "Endpoints for managing orders")
+@RestController
+@RequestMapping("/api/orders")
+@RequiredArgsConstructor
+public class OrderController {
```
Comment (karma-o) [inline src/main/java/com/example/bookshop/controller/OrderController.java:29]: Add roles who can access those endpoints according to the task description

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+        return orderItemMapper.toDto(orderItem);
+    }
+
+    @Override
+    public OrderResponseDto createNewOrder(User user, OrderRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getCartByUserId(user.getId())
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't found user with this id: "
+                        + user.getId())
+        );
```
Comment (karma-o) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java:89]: we should clear the cart items from the cart after we have placed the order

File: src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+        Order order = createOrder(user, requestDto);
+        Set<OrderItem> orderItems = getOrderItems(shoppingCart, order);
+        BigDecimal total = orderItems.stream()
+                .map(e -> e.getPrice()
+                        .multiply(BigDecimal.valueOf(e.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
```
Comment (karma-o) [inline src/main/java/com/example/bookshop/service/impl/OrderServiceImpl.java]: let's calculate the total price in a separate method

**Review by karma-o — APPROVED** [review]
GJ! take a look at a few of my comments as well

**Pull request merged**